README.md
# Logux Vuex

<img align="right" width="95" height="95" title="Logux logo"
     src="https://cdn.rawgit.com/logux/logux/master/logo.svg">

[![Build Status](https://travis-ci.org/nikolay-govorov/logux-vuex.svg?branch=master)](https://travis-ci.org/nikolay-govorov/logux-vuex)

Logux is a client-server communication protocol. It synchronizes action
between clients and server logs.

This library provides Vuex compatible API.

## Install

```sh
npm i --save logux-vuex
```

## Usage

Create Vuex store by `createLoguxStore`. It returns original Vuex store `Vuex.Store` function with Logux inside

```diff
import Vue from 'vue';
import Vuex from 'vuex';

+import createLoguxStore from 'logux-vuex/create-logux-store';

Vue.use(Vuex);

-const Store = Vuex.Store;
+const Store = createLoguxStore({
+  subprotocol: '1.0.0',
+  server: 'wss://localhost:1337',
+  userId: 10
+});

const store = new Store({
  state: {
    count: 0
  },
  mutations: {
    increment(state) {
      state.count++
    }
  }
})

+store.client.start()
```
See also [basic usage example](https://github.com/nikolay-govorov/logux-vuex-example) and [Logux Status] for UX best practices.

[Logux Status]: https://github.com/logux/logux-status

## Commit

Instead of Vuex, in Logux Vuex you have 4 ways to commit action:

* `store.commit(action)` is legacy API. Try to avoid it since you can’t
  specify how clean this actions.
* `store.commit.local(action, meta)` — action will be visible only to current
  browser tab.
* `store.commit.crossTab(action, meta)` — action will be visible
  to all browser tab.
* `store.commit.sync(action, meta)` — action will be visible to server
  and all browser tabs.

In all 3 new commit methods you must to specify `meta.reasons` with array
of “reasons”. It is code names of reasons, why this action should be still
in the log.

```js
store.commit.crossTab(
  { type: 'CHANGE_NAME', name }, { reasons: ['lastName'] }
)
```

When you don’t need some actions, you can remove reasons from them:

```js
store.commit.crossTab(
  { type: 'CHANGE_NAME', name }, { reasons: ['lastName'] }
).then(meta => {
  store.log.removeReason('lastName', { maxAdded: meta.added - 1 })
})
```

Action with empty reasons will be removed from log.


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.0.2
* Add the classical syntax of commits

## 0.1.1
* Generates JSDoc documentation
* Don’t apply action if it was deleted during waiting for replay end.

## 0.1
* Initial release.


test/index.test.js
var createLoguxStore = require('../create-logux-store')
var index = require('../')

it('has createLoguxCreator function', function () {
  expect(index.createLoguxStore).toBe(createLoguxStore)
})


test/create-logux-store.test.js
var Vue = require('vue')
var Vuex = require('vuex')

var TestPair = require('logux-sync').TestPair
var TestTime = require('logux-core').TestTime

Vue.use(Vuex)

var createLoguxStore = require('../create-logux-store')

function createStore (mutations, opts) {
  if (!opts) opts = { }
  if (!opts.server) opts.server = 'wss://localhost:1337'

  opts.subprotocol = '1.0.0'
  opts.userId = 10
  opts.time = new TestTime()

  var LoguxStore = createLoguxStore(opts)
  var store = new LoguxStore({
    state: { value: 0 },

    mutations: mutations
  })

  var prev = 0
  store.log.generateId = function () {
    prev += 1
    return [prev, store.client.options.userId + ':uuid', 0]
  }

  return store
}

function increment (state) {
  state.value = state.value + 1
}

function historyLine (state, action) {
  state.value = state.value + action.value
}

function actions (log) {
  return log.store.created.map(function (i) {
    return i[0]
  })
}

var originWarn = console.warn
afterEach(function () {
  console.warn = originWarn
})

it('throws error on missed config', function () {
  expect(function () {
    createLoguxStore()
  }).toThrowError('Missed server option in Logux client')
})

it('creates Vuex store', function () {
  var store = createStore({ increment: increment })

  store.commit({
    type: 'increment'
  })

  expect(store.state).toEqual({ value: 1 })
})

it('not Object type action', function () {
  var store = createStore({
    increment: function (state, count) {
      state.value = state.value + count
    }
  })

  store.commit('increment', 5)

  expect(store.state).toEqual({ value: 5 })
})

it('creates Logux client', function () {
  var store = createStore({ increment: increment })

  expect(store.client.options.subprotocol).toEqual('1.0.0')
})

it('sets tab ID', function () {
  var store = createStore({ increment: increment })

  return new Promise(function (resolve) {
    store.log.on('add', function (action, meta) {
      expect(meta.tab).toEqual(store.client.id)
      expect(meta.reasons).toEqual(['tab' + store.client.id])

      resolve()
    })

    store.commit({
      type: 'increment'
    })
  })
})

it('has shortcut for add', function () {
  var store = createStore({ increment: increment })

  return store.commit.crossTab(
    { type: 'increment' }, { reasons: ['test'] }
  ).then(function () {
    expect(store.state).toEqual({ value: 1 })
    expect(store.log.store.created[0][1].reasons).toEqual(['test'])
  })
})

it('listen for action from other tabs', function () {
  var store = createStore({ increment: increment })

  store.client.emitter.emit('add', { type: 'increment' }, { id: [1, 't', 0] })

  expect(store.state).toEqual({ value: 1 })
})

it('saves previous states', function () {
  var calls = 0

  var store = createStore({
    a: function () {
      calls += 1
    }
  })

  var promise = Promise.resolve()

  for (var i = 0; i < 60; i++) {
    if (i % 2 === 0) {
      promise = promise.then(function () {
        return store.commit.crossTab({ type: 'a' }, { reasons: ['test'] })
      })
    } else {
      store.commit({ type: 'a' })
    }
  }

  return promise.then(function () {
    expect(calls).toEqual(60)
    calls = 0

    return store.commit.crossTab(
      { type: 'a' }, { id: [57, '10:uuid', 1], reasons: ['test'] }
    )
  }).then(function () {
    expect(calls).toEqual(10)
  })
})

it('changes history recording frequency', function () {
  var calls = 0

  var store = createStore({
    a: function () {
      calls += 1
    }
  }, {
    saveStateEvery: 1
  })

  return Promise.all([
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] })
  ]).then(function () {
    calls = 0
    return store.commit.crossTab(
      { type: 'a' }, { id: [3, '10:uuid', 1], reasons: ['test'] })
  }).then(function () {
    expect(calls).toEqual(2)
  })
})

it('cleans its history on removing action', function () {
  var calls = 0
  var store = createStore({
    a: function () {
      calls += 1
    }
  }, {
    saveStateEvery: 2
  })

  return Promise.all([
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab({ type: 'a' }, { reasons: ['test'] })
  ]).then(function () {
    return store.log.changeMeta([5, '10:uuid', 0], { reasons: [] })
  }).then(function () {
    calls = 0
    return store.commit.crossTab(
      { type: 'a' }, { id: [5, '10:uuid', 1], reasons: ['test'] })
  }).then(function () {
    expect(calls).toEqual(3)
  })
})

it('changes history', function () {
  var store = createStore({ historyLine: historyLine })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['test'] }
    ),
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }
    )
  ]).then(function () {
    store.commit({ type: 'historyLine', value: 'c' })
    store.commit({ type: 'historyLine', value: 'd' })

    return store.commit.crossTab(
      { type: 'historyLine', value: '|' },
      { id: [2, '10:uuid', 1], reasons: ['test'] }
    )
  }).then(function () {
    expect(store.state.value).toEqual('0ab|cd')
  })
})

it('undoes actions', function () {
  var store = createStore({ historyLine: historyLine })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'c' }, { reasons: ['test'] })
  ]).then(function () {
    expect(store.state.value).toEqual('0abc')

    return store.commit.crossTab(
      { type: 'logux/undo', id: [2, '10:uuid', 0] }, { reasons: ['test'] }
    )
  }).then(function () {
    expect(store.state.value).toEqual('0ac')
  })
})

it('warns about undoes cleaned action', function () {
  console.warn = jest.fn()
  var store = createStore({ increment: increment })

  return store.commit.crossTab(
    { type: 'logux/undo', id: [1, 't', 0] }, { reasons: [] }
  ).then(function () {
    expect(console.warn).toHaveBeenCalledWith(
      'Logux can not find [1,"t",0] to undo it. Maybe action was cleaned.'
    )
  })
})

it('replays history since last state', function () {
  var onMissedHistory = jest.fn()

  var store = createStore({ historyLine: historyLine }, {
    onMissedHistory: onMissedHistory,
    saveStateEvery: 2
  })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['one'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'c' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'd' }, { reasons: ['test'] })
  ]).then(function () {
    return store.log.removeReason('one')
  }).then(function () {
    return store.commit.crossTab(
      { type: 'historyLine', value: '|' },
      { id: [1, '10:uuid', 0], reasons: ['test'] }
    )
  }).then(function () {
    expect(onMissedHistory)
      .toHaveBeenCalledWith({ type: 'historyLine', value: '|' })
    expect(store.state.value).toEqual('0abc|d')
  })
})

it('replays history for reason-less action', function () {
  var store = createStore({ historyLine: historyLine })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'c' }, { reasons: ['test'] })
  ]).then(function () {
    return store.commit.crossTab(
      { type: 'historyLine', value: '|' }, { id: [1, '10:uuid', 1] }
    )
  }).then(function () {
    return Promise.resolve()
  }).then(function () {
    expect(store.state.value).toEqual('0a|bc')
    expect(store.log.store.created).toHaveLength(3)
  })
})

it('replays actions before staring since initial state', function () {
  var onMissedHistory = jest.fn()
  var store = createStore({ historyLine: historyLine }, {
    onMissedHistory: onMissedHistory,
    saveStateEvery: 2
  })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'c' }, { reasons: ['test'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'd' }, { reasons: ['test'] })
  ]).then(function () {
    return store.commit.crossTab(
      { type: 'historyLine', value: '|' },
      { id: [0, '10:uuid', 0], reasons: ['test'] }
    )
  }).then(function () {
    expect(store.state.value).toEqual('0|bcd')
  })
})

it('replays actions on missed history', function () {
  var onMissedHistory = jest.fn()

  var store = createStore({ historyLine: historyLine }, {
    onMissedHistory: onMissedHistory
  })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['one'] }),
    store.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] })
  ]).then(function () {
    return store.log.removeReason('one')
  }).then(function () {
    return store.commit.crossTab(
      { type: 'historyLine', value: '|' },
      { id: [0, '10:uuid', 0], reasons: ['test'] }
    )
  }).then(function () {
    expect(onMissedHistory)
      .toHaveBeenCalledWith({ type: 'historyLine', value: '|' })
    expect(store.state.value).toEqual('0|b')
  })
})

it('does not fall on missed onMissedHistory', function () {
  var store = createStore({ historyLine: historyLine })

  return Promise.all([
    store.commit.crossTab(
      { type: 'historyLine', value: 'a' }, { reasons: ['first'] })
  ]).then(function () {
    return store.log.removeReason('first')
  }).then(function () {
    return store.commit.crossTab(
      { type: 'historyLine', value: '|' },
      { id: [0, '10:uuid', 0], reasons: ['test'] }
    )
  }).then(function () {
    expect(store.state.value).toEqual('0|')
  })
})

it('cleans action added by commit', function () {
// TODO: fix commitHistory

//   var store = createStore({ historyLine: historyLine }, {
//     commitHistory: 3
//   })

//   function add (index) {
//     return function () {
//       store.commit({ type: 'historyLine', value: index })
//     }
//   }

//   var promise = Promise.resolve()
//   for (var i = 1; i <= 25; i++) {
//     promise = promise.then(add(i))
//   }

//   return promise.then(function () {
//     expect(actions(store.log)).toEqual([
//       { type: 'historyLine', value: 25 },
//       { type: 'historyLine', value: 24 },
//       { type: 'historyLine', value: 23 }
//     ])
//   })
})

it('cleans last 1000 by default', function () {
  var store = createStore({ increment: increment })

  var promise = Promise.resolve()

  for (var i = 0; i < 1050; i++) {
    promise = promise.then(function () {
      store.commit({ type: 'increment' })
    })
  }

  return promise.then(function () {
    expect(actions(store.log)).toHaveLength(1000)
  })
})

it('copies reasons to undo action', function () {
  var store = createStore({ increment: increment })

  return store.commit.crossTab(
    { type: 'increment' }, { reasons: ['a', 'b'] }
  ).then(function () {
    return store.commit.crossTab(
      { type: 'logux/undo', id: [1, '10:uuid', 0] }, { reasons: [] })
  }).then(function () {
    return store.log.byId([2, '10:uuid', 0])
  }).then(function (result) {
    expect(result[0].type).toEqual('logux/undo')
    expect(result[1].reasons).toEqual(['a', 'b'])
  })
})

it('does not override undo action reasons', function () {
  var store = createStore({ increment: increment })

  return store.commit.crossTab(
    { type: 'increment' }, { reasons: ['a', 'b'] }
  ).then(function () {
    return store.commit.crossTab(
      { type: 'logux/undo', id: [1, '10:uuid', 0] },
      { reasons: ['c'] }
    )
  }).then(function () {
    return store.log.byId([2, '10:uuid', 0])
  }).then(function (result) {
    expect(result[0].type).toEqual('logux/undo')
    expect(result[1].reasons).toEqual(['c'])
  })
})

it('commites local actions', function () {
  var store = createStore({ increment: increment })

  return store.commit.local(
    { type: 'increment' }, { reasons: ['test'] }
  ).then(function () {
    expect(store.log.store.created[0][0]).toEqual({ type: 'increment' })
    expect(store.log.store.created[0][1].tab).toEqual(store.client.id)
    expect(store.log.store.created[0][1].reasons).toEqual(['test'])
  })
})

it('commites sync actions', function () {
  var store = createStore({ increment: increment })

  return store.commit.sync(
    { type: 'increment' }, { reasons: ['test'] }
  ).then(function () {
    var log = store.log.store.created

    expect(log[0][0]).toEqual({ type: 'increment' })
    expect(log[0][1].sync).toBeTruthy()
    expect(log[0][1].reasons).toEqual(['test', 'waitForSync'])
  })
})

it('cleans sync action after synchronization', function () {
  var pair = new TestPair()
  var store = createStore({ increment: increment }, { server: pair.left })

  store.client.start()
  return pair.wait('left').then(function () {
    var protocol = store.client.sync.localProtocol
    pair.right.send(['connected', protocol, 'server', [0, 0]])
    return store.client.sync.waitFor('synchronized')
  }).then(function () {
    store.commit.sync({ type: 'increment' })
    return pair.wait('right')
  }).then(function () {
    expect(actions(store.log)).toEqual([{ type: 'increment' }])
    pair.right.send(['synced', 1])
    return store.client.sync.waitFor('synchronized')
  }).then(function () {
    return Promise.resolve()
  }).then(function () {
    expect(actions(store.log)).toEqual([])
  })
})

it('applies old actions from store', function () {
  var store1 = createStore({ historyLine: historyLine })
  var store2

  return Promise.all([
    store1.commit.crossTab(
      { type: 'historyLine', value: '1' },
      { id: [0, '10:x', 1], reasons: ['test'] }
    ),
    store1.commit.crossTab(
      { type: 'historyLine', value: '2' },
      { id: [0, '10:x', 2], reasons: ['test'] }
    ),
    store1.commit.crossTab(
      { type: 'historyLine', value: '3' },
      { id: [0, '10:x', 3], reasons: ['test'] }
    ),
    store1.commit.crossTab(
      { type: 'historyLine', value: '4' },
      { id: [0, '10:x', 4], reasons: ['test'] }
    ),
    store1.log.add(
      { type: 'historyLine', value: '5' },
      { id: [0, '10:x', 5], reasons: ['test'], tab: store1.client.id }
    ),
    store1.commit.crossTab(
      { type: 'logux/undo', id: [0, '10:x', 2] },
      { id: [0, '10:x', 6], reasons: ['test'] }
    )
  ]).then(function () {
    store2 = createStore(
      { historyLine: historyLine }, { store: store1.log.store })

    store2.commit({ type: 'historyLine', value: 'a' })
    store2.commit.crossTab(
      { type: 'historyLine', value: 'b' }, { reasons: ['test'] }
    )
    expect(store2.state.value).toEqual('0a')

    return store2.initialize
  }).then(function () {
    expect(store2.state.value).toEqual('0134ab')
  })
})


create-logux-store.js
var Store = require('vuex').Store
var nanoclone = require('nanoclone')
var isFirstOlder = require('logux-core/is-first-older')
var CrossTabClient = require('logux-client/cross-tab-client')

function warnBadUndo (id) {
  var json = JSON.stringify(id)

  console.warn(
    'Logux can not find ' + json + ' to undo it. Maybe action was cleaned.'
  )
}

function LoguxState (client, config, vuexConfig) {
  var self = this

  Store.call(this, nanoclone(vuexConfig))

  this.client = client
  this.log = client.log

  var init
  var prevMeta
  var replaying
  var wait = {}
  var history = {}
  var actionCount = 0
  var started = (function (now) {
    return { id: now, time: now[0] }
  })(client.log.generateId())

  self.initialize = new Promise(function (resolve) {
    init = resolve
  })

  self.commit = function commit () {
    var action
    var isNoObjectTypeAction

    if (typeof arguments[0] === 'string') {
      var type = arguments[0]
      var options = Array.from(arguments).slice(1)
      isNoObjectTypeAction = true

      action = {
        type: type,
        options: options
      }
    } else {
      action = arguments[0]
    }

    var meta = {
      id: client.log.generateId(),
      tab: client.id,
      reasons: ['tab' + client.id],
      dispatch: true
    }

    client.log.add(action, meta)

    prevMeta = meta
    originCommit(action, isNoObjectTypeAction)
    saveHistory(meta)
  }

  self.commit.local = function local (action, meta) {
    meta.tab = client.id
    return client.log.add(action, meta)
  }

  self.commit.crossTab = function crossTab (action, meta) {
    return client.log.add(action, meta)
  }

  self.commit.sync = function sync (action, meta) {
    if (!meta) meta = {}
    if (!meta.reasons) meta.reasons = []

    meta.sync = true
    meta.reasons.push('waitForSync')

    return client.log.add(action, meta)
  }

  client.log.on('preadd', function (action, meta) {
    if (action.type === 'logux/undo' && meta.reasons.length === 0) {
      meta.reasons.push('reasonsLoading')
    }

    if (!isFirstOlder(prevMeta, meta) && meta.reasons.length === 0) {
      meta.reasons.push('replay')
    }
  })

  var lastAdded = 0
  var dispatchCalls = 0
  client.on('add', function (action, meta) {
    if (meta.added > lastAdded) lastAdded = meta.added

    if (meta.dispatch) {
      dispatchCalls += 1

      if (lastAdded > config.dispatchHistory && dispatchCalls % 25 === 0) {
        client.log.removeReason('tab' + client.id, {
          maxAdded: lastAdded - config.dispatchHistory
        })
      }

      return
    }

    process(action, meta, isFirstOlder(meta, started))
  })

  client.on('clean', function (action, meta) {
    var key = meta.id.join('\t')

    delete wait[key]
    delete history[key]
  })

  client.sync.on('state', function () {
    if (client.sync.state === 'synchronized') {
      client.log.removeReason('waitForSync', { maxAdded: client.sync.lastSent })
    }
  })

  var previous = []
  var ignores = {}
  client.log.each(function (action, meta) {
    if (!meta.tab) {
      if (action.type === 'logux/undo') {
        ignores[action.id.join('\t')] = true
      } else if (!ignores[meta.id.join('\t')]) {
        previous.push([action, meta])
      }
    }
  }).then(function () {
    if (previous.length > 0) {
      previous.forEach(function (i) {
        process(i[0], i[1], true)
      })
    }

    init()
  })

  // Functions
  function saveHistory (meta) {
    actionCount += 1

    if (
      config.saveStateEvery === 1 || actionCount % config.saveStateEvery === 1
    ) {
      history[meta.id.join('\t')] = nanoclone(self.state)
    }
  }

  function originCommit (action, isNotObjectTypeAction) {
    if (action.type === 'logux/state') {
      self.replaceState(action.state)

      return
    }

    var commitArgs = arguments

    if (isNotObjectTypeAction) {
      commitArgs = [action.type].concat(action.options)
    }

    if (action.type in self._mutations) {
      Store.prototype.commit.apply(self, commitArgs)
    }
  }

  function replaceState (state, actions) {
    var newState = actions.reduceRight(function (prev, i) {
      var changed = nanoclone(prev)

      if (vuexConfig.mutations[i[0].type]) {
        vuexConfig.mutations[i[0].type](changed, i[0])
      }

      if (history[i[1]]) history[i[1]] = changed

      return changed
    }, state)

    originCommit({ type: 'logux/state', state: newState })
  }

  function replay (actionId, replayIsSafe) {
    var until = actionId.join('\t')

    var ignore = {}
    var actions = []
    var replayed = false
    var newAction
    var collecting = true

    replaying = new Promise(function (resolve) {
      client.log.each(function (action, meta) {
        if (meta.tab && meta.tab !== client.id) return true

        var id = meta.id.join('\t')

        if (collecting || !history[id]) {
          if (action.type === 'logux/undo') {
            ignore[action.id.join('\t')] = true
            return true
          }

          if (!ignore[id]) actions.push([action, id])
          if (id === until) {
            newAction = action
            collecting = false
          }

          return true
        } else {
          replayed = true
          replaceState(history[id], actions)

          return false
        }
      }).then(function () {
        if (!replayed) {
          if (config.onMissedHistory) config.onMissedHistory(newAction)

          var full
          if (replayIsSafe) {
            full = actions
          } else {
            full = actions.slice(0)
            while (actions.length > 0) {
              var last = actions[actions.length - 1]
              actions.pop()
              if (history[last[1]]) {
                replayed = true
                replaceState(history[last[1]], actions.concat([
                  [newAction, until]
                ]))
                break
              }
            }
          }

          if (!replayed) {
            replaceState(nanoclone(vuexConfig.state), full)
          }
        }

        replaying = false
        resolve()
      })
    })

    return replaying
  }

  function process (action, meta, replayIsSafe) {
    if (replaying) {
      var key = meta.id.join('\t')
      wait[key] = true

      replaying.then(function () {
        if (wait[key]) {
          process(action, meta, replayIsSafe)

          delete wait[key]
        }
      })

      return
    }

    if (action.type === 'logux/undo') {
      var reasons = meta.reasons
      client.log.byId(action.id).then(function (result) {
        if (result[0]) {
          if (reasons.length === 1 && reasons[0] === 'reasonsLoading') {
            client.log.changeMeta(meta.id, { reasons: result[1].reasons })
          }
          delete history[action.id.join('\t')]
          replay(action.id)
        } else {
          client.log.changeMeta(meta.id, { reasons: [] })
          warnBadUndo(action.id)
        }
      })
    } else if (isFirstOlder(prevMeta, meta)) {
      prevMeta = meta
      originCommit(action)
      if (meta.added) saveHistory(meta)
    } else {
      replay(meta.id, replayIsSafe).then(function () {
        if (meta.reasons.indexOf('replay') !== -1) {
          client.log.changeMeta(meta.id, {
            reasons: meta.reasons.filter(function (i) {
              return i !== 'replay'
            })
          })
        }
      })
    }
  }
}

LoguxState.prototype = Object.create(Store.prototype)

LoguxState.prototype.constructor = LoguxState

function createLoguxState (config) {
  if (!config) config = {}

  var client = new CrossTabClient(config)

  return LoguxState.bind(null, client, {
    dispatchHistory: config.dispatchHistory || 1000,
    saveStateEvery: config.saveStateEvery || 50,
    onMissedHistory: config.onMissedHistory
  })
}

module.exports = createLoguxState
